File: form/schema.py
"""
Schema module responsible for transforming raw form data into the structured format
"""
 
 
from datetime import datetime, timezone
import os
import requests
from dotenv import load_dotenv  
load_dotenv()
 
MAIN_TABLE_NAME = "STUDENT_DATA"
CLASS_TABLE_NAME = "CLASSIFICATION"
SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_KEY")
 
def schema(form_data: dict):
    created_at = datetime.now(timezone.utc).isoformat()
 
    # Build the STUDENT_DATA row.
    latest_id_url = (
        f"{SUPABASE_URL}/rest/v1/{MAIN_TABLE_NAME}"
        f"?select=student_id&order=student_id.desc&limit=1"
    )
    latest_id_res = requests.get(
        latest_id_url,
        headers={
            "apikey": SUPABASE_KEY,
            "Authorization": f"Bearer {SUPABASE_KEY}",
            "Accept": "application/json",
        },
        timeout=30,
    )
    latest_student = latest_id_res.json() if latest_id_res.text else []
    latest_student_id = None
    if isinstance(latest_student, list) and latest_student:
        latest_student_id = latest_student[0].get("student_id")
 
    try:
        latest_student_id_int = int(latest_student_id) if latest_student_id not in (None, "") else 0
    except Exception:
        latest_student_id_int = 0
 
    global new_student_id
    new_student_id = latest_student_id_int + 1
    
    data = {
        'student_id': new_student_id,
        'created_at': created_at,
        'student_name': form_data.get('section0', {}).get('student_name', ''),
        'mobile_number': form_data.get('section0', {}).get('mobile_number', ''),
        'email_id': form_data.get('section0', {}).get('email_id', ''),
        'college_name': form_data.get('section0', {}).get('college_name', ''),
        'department': form_data.get('section0', {}).get('department', ''),
        'course_type': form_data.get('section1', {}).get('course_type', ''),
        'year': str(form_data.get('section1', {}).get('year', '')),
        'medium': form_data.get('section1', {}).get('medium', ''),
        'have_prep_test': form_data.get('section1', {}).get('have_prep_test', ''),
        'career_goal': form_data.get('section1', {}).get('career_goal', ''),
    }
 
    s2 = form_data.get('section2', {})
    for subj in ['quant_answers', 'logic_answers', 'verbal_answers']:
        answers = s2.get(subj, [])
        prefix = subj.split('_')[0]
        for i in range(5):
            data[f'{prefix}_Q{i+1}'] = str(answers[i]) if i < len(answers) else ''
 
    s3 = form_data.get('section3', {})
    for i in range(1, 6):
        data[f'behave_Q{i}'] = str(s3.get(f'behave_Q{i}', ''))
 
    s4 = form_data.get('section4', {})
    for i in range(1, 5):
        data[f'learn_Q{i}'] = str(s4.get(f'learn_Q{i}', ''))
 
    s5 = form_data.get('section5', {})
    data.update({
        'instruct_Q1': str(s5.get('instruct_Q1', '')),
        'instruct_Q2': str(s5.get('instruct_Q2', '')),
        'instruct_Q3': str(s5.get('instruct_Q3', '')),
        'instruct_Q4': str(s5.get('instruct_Q4', '')),
        'instruct_Q5': str(s5.get('instruct_Q5', '')),
        'instruct_Q6': str(s5.get('instruct_Q6', '')),
    })
 
    s5a = form_data.get('section5a', {})
    data['content_pref_Q1'] = str(s5a.get('content_pref_Q1', s5a.get('content_Q1', '')))
    data['content_pref_Q2'] = str(s5a.get('content_pref_Q2', s5a.get('content_Q2', '')))
    data['content_pref_Q3'] = str(s5a.get('content_pref_Q3', s5a.get('content_Q3', '')))
 
    s5b = form_data.get('section5b', {})
    data.update({
        'engage_Q1': str(s5b.get('engage_Q1', '')),
        'engage_Q2': str(s5b.get('engage_Q2', '')),
        'engage_Q3': str(s5b.get('engage_Q3', '')),
        'engage_Q4': str(s5b.get('engage_Q4', '')),
        'commit_Q1': str(s5b.get('commit_Q1', '')),
        'commit_Q2': str(s5b.get('commit_Q2', '')),
        'commit_Q3': str(s5b.get('commit_Q3', '')),
        'commit_Q4': str(s5b.get('commit_Q4', '')),
    })
    return data
 

File: frontend/js/schema.js
/**
 * schema.js
 * Builds the STUDENT_DATA row object from the raw form sections.
 */
 
export function buildStudentRow(formData) {
    const s0  = formData.section0  || {};
    const s1  = formData.section1  || {};
    const s2  = formData.section2  || {};
    const s3  = formData.section3  || {};
    const s4  = formData.section4  || {};
    const s5  = formData.section5  || {};
    const s5a = formData.section5a || {};
    const s5b = formData.section5b || {};

    const row = {
        created_at:     new Date().toISOString(),

        student_id:     undefined,

        student_name:   s0.student_name   || '',
        mobile_number:  s0.mobile_number  || '',
        email_id:       s0.email_id       ?? null,
        college_name:   s0.college_name   || '',
        department:     s0.department     || '',

        course_type:    s1.course_type    || '',
        year:           s1.year != null ? String(s1.year) : '',
        medium:         s1.medium         || '',
        have_prep_test: s1.have_prep_test || '',
        career_goal:    s1.career_goal    || '',
    };
 
    for (const [subj, prefix] of [
        ['quant_answers',  'quant'],
        ['logic_answers',  'logic'],
        ['verbal_answers', 'verbal'],
    ]) {
        const answers = s2[subj] || [];
        for (let i = 0; i < 5; i++) {
            row[`${prefix}_Q${i + 1}`] = String(answers[i] ?? '');
        }
    }
 
    for (let i = 1; i <= 5; i++) {
        row[`behave_Q${i}`] = String(s3[`behave_Q${i}`] ?? '');
    }
 
    for (let i = 1; i <= 4; i++) {
        row[`learn_Q${i}`] = String(s4[`learn_Q${i}`] ?? '');
    }
 
    for (let i = 1; i <= 6; i++) {
        row[`instruct_Q${i}`] = String(s5[`instruct_Q${i}`] ?? '');
    }
 
    row['content_pref_Q1'] = String(s5a.content_pref_Q1 ?? s5a.content_Q1 ?? '');
    row['content_pref_Q2'] = String(s5a.content_pref_Q2 ?? s5a.content_Q2 ?? '');
    row['content_pref_Q3'] = String(s5a.content_pref_Q3 ?? s5a.content_Q3 ?? '');
 
    row['engage_Q1'] = String(s5b.engage_Q1  ?? '');
    row['engage_Q2'] = String(s5b.engage_Q2  ?? '');
    row['engage_Q3'] = String(s5b.engage_Q3  ?? '');
    row['engage_Q4'] = String(s5b.engage_Q4  ?? '');
    row['commit_Q1'] = String(s5b.commit_Q1  ?? '');
    row['commit_Q2'] = String(s5b.commit_Q2  ?? '');
    row['commit_Q3'] = String(s5b.commit_Q3  ?? '');
    row['commit_Q4'] = String(s5b.commit_Q4  ?? '');
 
    return row;
}